### DISCUSSION TITLE
D3D12: Vertex buffer input:  How to use any other slot (from 1 to 31), besides default 0 ?

### DISCUSSION THREAD

**Discussion by MarcioAB** [amerkoleci/Vortice.Windows#362] [Q&A] [answered]
Besides default slot 0, I want to use any other slot (from 1 to 31), to input a vertex buffer in D3D12 . For D3D11 it works ok but for D3D12 it only works for slot 0. No error but no render also.

Below are the 2 instructions, I'm using to set up the vertex buffer, both D3D11 & D3D12.
I must be missing something for D3D12, but I can not figure out what.

The InputLayout is the same for both D3D11 & D3D12: InputElementDescription("I", 0, format, offset, slot_number)

The IASetVertexBuffers is also very similar:
   D3D11: IASetVertexBuffer(slot_number, vBuffer, stride);
   D3D12: IASetVertexBuffers(slot_number, vBufferView);

Any help is much appreciated.

**Comment by amerkoleci** [comment]
Doesn't this method works?

https://github.com/amerkoleci/Vortice.Windows/blob/main/src/Vortice.Direct3D12/ID3D12GraphicsCommandList.cs#L588

**Reply by MarcioAB** [reply]
This is the method that I'm using.

But due to my lack of basic understanding of D3D12 it is difficult for me to say it is not working because of this method.
Considering the D3D11 working environment and also using your D3D12 Hello samples, I understand the necessary instructions for slot definition are only the Input Layout and in IASetVertexBuffer.
On the other side, considering slot 0 is the default, there could exist another set I'm missing that is necessary for any slot that is not the default 0. 
But if only those 2 instructions are the ones necessary, then maybe something could be wrong.

**Comment by amerkoleci** [comment]
You're sure you're reading from input element from vertex shader?

**Reply by MarcioAB** [reply]
Do not know exactly how to be sure, BUT it works fine when slot=0 and does not work anymore if I change it to slot=1.
All code related to this slot is based on a variable and not hard-coded.

The vertex shader is as simple as possible:

float4 VSMain(float2 position : I) : SV_Position
{ return mul(MVP, float4(position, 0, 1)); }

The input element for "I" is InputElementDescription("I", 0, format, offset, slot_number).
The InputElementDescription[] has only this element.

**Reply by MarcioAB** [reply]
Sorry. Just found my error.

I was using this same vertex slot in SetGraphicsRootDescriptorTable(slot_number, hMVP) but I just discovered I need to use 0 for this slot. So, for now, until I discover what kind of slot is, I need to use SetGraphicsRootDescriptorTable(0, hMVP) with this slot being hard-coded with 0 (zero).
hMVP is a handle for the constant buffer carrying on the MVP transformation.

**Comment by MarcioAB** [comment, marked as answer]
I was using this slot in SetGraphicsRootDescriptorTable(slot, hMVP) for a constant buffer.
SetGraphicsRootDescriptorTable uses a different kind of slot, so it will be hard-coded with zero for now.
